src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs

```diff
@@ -2013,30 +2036,40 @@ protected override void BeginProcessing()
                 }
             }
 
-            if (Wait.IsPresent)
+            if (Wait.IsPresent || _timeoutSpecified)
             {
                 if (process != null)
                 {
                     if (!process.HasExited)
                     {
 #if UNIX
-                        process.WaitForExit();
+                        if (!process.WaitForExit(_timeout))
+                        {
+                            StopProcessOnTimeout(process);
+                        }
 #else
                         _waithandle = new ManualResetEvent(false);
 
                         // Create and start the job object
                         ProcessCollection jobObject = new ProcessCollection();
                         if (jobObject.AssignProcessToJobObject(process))
                         {
-                            // Wait for the job object to finish
-                            jobObject.WaitOne(_waithandle);
+                            // Wait for the job object to finish, or kill it if a timeout occurs
+                            jobObject.WaitOne(_waithandle, _timeout);
+                            if (!process.HasExited)
+                            {
+                                StopProcessOnTimeout(process);
```

Comment:
I'm wondering if we actually want to stop the process on timeout by default.  I think it might be sufficient to return a non-terminating error so that `-PassThru` returns the process so the user can terminate it if so desired?  cc @HemantMahawar @joeyaiello

Reply:
From a user's perspective, I'd expect `PowerShell` to stop the process for me because this is how at least I would use this parameter.

Reply:
I wonder why we're actually making this change if the following works fine `Start-Process -PassThru | Wait-Process -Timeout`? It seems we duplicate a code.
I'd expect that we enhance Wait-Process cmdlet with -PassThru to get `Start-Process -PassThru | Wait-Process -Timeout 10 -PassThru | Stop-Process`.

Reply:
@SteveL-MSFT @bergmeister I thought about both of these ways when implementing this, but settled on stopping the process on timeout, similar to [https://ss64.com/bash/timeout.html](https://ss64.com/bash/timeout.html). The reasoning is if I started a process and gave it a time limit, I would expect it to stop if the limit is reached, rather than just let me know it had gone over the limit. Is it worth adding a switch `-StopOnExitTimeout` (or similar)? Or, should the below suggestion be used?

@iSazonov I see your point, and didn't think of it being run this way. I suppose adding the `-ExitTimeout` parameter is about simplifying the amount of cmdlets needed. Would you prefer me to implement `-PassThru` for `Wait-Process` instead?